blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/listeners/DownstreamJobListener.java

```diff
@@ -0,0 +1,130 @@
+ * work later to retrieve the details.
+ */
+@Extension
+public class DownstreamJobListener extends RunListener<Run<?, ?>> {
+
+    private static final Logger LOGGER = Logger.getLogger(DownstreamJobListener.class.getName());
+
+    @Override
+    public void onStarted(Run<?, ?> run, TaskListener listener) {
+
+        for (BuildUpstreamNodeAction action : run.getActions(BuildUpstreamNodeAction.class)) {
+            Run triggerRun = Run.fromExternalizableId(action.getUpstreamRunId());
+            if (triggerRun instanceof WorkflowRun) {
+                FlowExecution execution = ((WorkflowRun) triggerRun).getExecution();
+                FlowNode node;
+
+                if (execution == null) {
+                    LOGGER.warning("Could not retrieve upstream FlowExecution");
+                    continue;
+                }
+
+                try {
+                    node = execution.getNode(action.getUpstreamNodeId());
+                } catch (IOException e) {
+                    LOGGER.warning("Could not retrieve upstream node: " + e);
+                    continue;
+                }
+
+                if (node == null) {
+                    LOGGER.warning("Could not retrieve upstream node (null)");
+                    continue;
+                }
+
+                // Add an action on the triggerRun node pointing to the currently executing run
+                String description = run.getDescription();
+                if (description == null) {
+                    description = run.getFullDisplayName();
+                }
+
+                // Find the node that will be used to create the cut-down graph for visualisation
+                FlowNode blueGraphNode = findBlueGraphNode(node);
+
+                if (blueGraphNode == null) {
+                    LOGGER.warning("Could not find a suitable parent node for node " + node);
+                    continue;
+                }
+
+                Link link = LinkResolver.resolveLink(run);
+                if (link != null) {
+                    try {
+                           // Add to the node we'll use for the visualisation
+                        blueGraphNode.addAction(new NodeDownstreamBuildAction(link, description));
+                        blueGraphNode.save();
+
+                        if (!blueGraphNode.equals(node)) {
+                            // Also add to the actual trigger node so we can find it later by step
+                            node.addAction(new NodeDownstreamBuildAction(link, description));
+                            node.save();
+                        }
+
+                    } catch (IOException e) {
+                        LOGGER.severe("Could not persist node: " + e);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Walks the execution graph from the originating node for a downstream build (actionNode), looking for the
```

Comment:
Looks like you are finding closest parallel or stage. What about the case where closest parallel or stage is nested? pipeline graph in UI shoes only top level parallel or stage. Given this, I think this should be looking for farthest parallel or stage?

Reply:
To be honest I can't think of a way to do this without re-implementing most of `PipelineNodeGraphVisitor` which is why I wrote the version that collected the actions while building the graph itself.

Reply:
@sophistifunk I have taken crack at a utility that gives you enclosing top level stage or parallel. See https://github.com/jenkinsci/blueocean-plugin/blob/1313ec172b180cdae71f3494ad6b1dc4dde300c2/blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java#L217.

There are tests to demonstrate nested, orphan cases working. PTAL. If it looks good you can merge PR #1626 in to this and we can close that.